cabal-install/Distribution/Client/Dependency.hs

```diff
@@ -587,13 +587,12 @@ interpretPackagesPreference :: Set PackageName
                             -> (PackageName -> PackagePreferences)
 interpretPackagesPreference selected defaultPref prefs =
   \pkgname -> PackagePreferences (versionPref pkgname) (installPref pkgname)
-
   where
     versionPref pkgname =
-      fromMaybe anyVersion (Map.lookup pkgname versionPrefs)
-    versionPrefs = Map.fromList
-      [ (pkgname, pref)
-      | PackageVersionPreference pkgname pref <- prefs ]
+      fromMaybe [anyVersion] (Map.lookup pkgname versionPrefs)
+    versionPrefs = foldr (\(PackageVersionPreference pkgname pref) m ->
+                            Map.insertWith (++) pkgname [pref] m)
+                   Map.empty prefs
```

Comment:
This is a partial function now, and it could fail in a sandbox.  You could use a list comprehension and `Map.fromListWith`, since a pattern in a list comprehension acts as a filter.